### ISSUE TITLE
External definitions with $ref JSON-pointer

### ISSUE THREAD

**Issue by vmanasiev** [json-editor/json-editor#379]
### General information
I want to reuse external enumerations for several main schemas.
There is a problem referencing external definitions.
According to the schema documentation we should be able to reference external definitions
with reference like this :       **"$ref" : "[url]#/definitions/[type]"**
As pointed in several sites like this:
http://www.tothenew.com/blog/using-definitions-internal-to-json-schema-and-external-to-json-schema-version-draft-v4/

- json-editor version: 1.3.5

### Expected behavior
To have a drop-down with the values defined in the external definition when referencing another json at the same level as the main document with : 
_"$ref": "./enums.json#/definitions/types"
"$ref":"enums.json#/definitions/types"_
### Actual behavior
The enums.json file is actually found and fetched with an ajax call.  
The field is broken and displays a drop-down with property types (string, object, etc..)
### Steps to reproduce the behavior

Two schema files defined in the same project folder

main.json
```javascript
{  
   "$schema":"http://json-schema.org/draft-04/schema#",
   "type":"object",
   "definitions":{  
      "envInternal":{  
         "type":"string",
         "enum":[  
            "Production",
            "UAT",
            "Performance",
            "Development",
            "Test"
         ]
      }
   },
   "properties":{  
      "brokenDropDown":{  
         "title":"Environment",
         "$ref":"enums.json#/definitions/envExternal"
      },
  "workingDropDown":{  
         "title":"Environment",
         "$ref":"#/definitions/envInternal"
      }
   }
}
```

enums.json
```javascript
{  
   "$schema":"http://json-schema.org/draft-04/schema#",
   "definitions":{  
      "envExternal":{  
         "type":"string",
         "enum":[  
            "Production",
            "UAT",
            "Performance",
            "Development",
            "Test"
         ]
      }
   }
}
```

**Comment by pmk65** [comment]
Maybe related to #156?

**Comment by vmanasiev** [comment]
@pmk I don't think it is the same issue. 

So it comes down of how we treat this "complex" references. 
when you have
```javascript
"$ref":"enums.json#/definitions/envExternal"
```
The ajax call succeeds and the enums.json file is loaded because the server does not mind what comes after #.
Then the code start to build a reference dictionaries and it stores the whole reference which leads to some problems.
So what I did was to treat  the reference in two parts
1)the part before # is used to load the enum.json file and is put in the reference dictionary by its own
2)the part after # is stored like other local reference but with updated  URL where the enum.json is added the path
I had to get rid off   "./" in front of the relative references because for some reason if we have several nested references it starts to build  the url like this   xxxx/./././enum.json which after that is not treated as  xxxx/./enum.json.
So I will put the code modifications in this comment , but there are also other things to fix.

In function  _getExternalRef
```javascript
 if (schema.$ref && typeof schema.$ref !== "object") {
        if (schema.$ref.substr(0,2)==="./") schema.$ref=schema.$ref.substr(2); //not a good place to do this.
        var ref_counter = this.refs_prefix + this.refs_counter++;
        if (schema.$ref.substr(0, 1) !== "#" && !this.refs[schema.$ref]) {
          var urlSplit = schema.$ref.split("#");
          refs[urlSplit[0]] = true;
          this.refs_with_info[ref_counter] = { fetchUrl: fetchUrl, "$ref": urlSplit[0] };
          if (urlSplit.length > 1) {
            ref_counter = this.refs_prefix + this.refs_counter++
            this.refs_with_info[ref_counter] = { fetchUrl: this._getFileBaseFromFileLocation(fetchUrl) + urlSplit[0], "$ref": "#" + urlSplit[1] };
          }
        } else {
          this.refs_with_info[ref_counter] = { fetchUrl: fetchUrl, "$ref": schema.$ref };
        }
        schema.$ref = ref_counter;
      }
```

**Comment by pmk65** [comment]
@vmanasiev I saw a similar issue when I looked at the old jdorn repo. https://github.com/jdorn/json-editor/pull/720
This looks very similar to your problem, and @desertkun provided a patch too.

**Comment by vmanasiev** [comment]
@pmk65 
To some extend it seems to address the same issue. Although this change is made on top of an older code and I don't think it is applicable straight away.
So when I have time I will try to push my modifications for a review.
I think I have addressed two more issues :  
1) work correctly with "up-a-level" relative paths like  "../xxxxxx"
2) Managing correctly resources with same name
 Lets say we have to json files with same name "/path1/fileName.json" and "/path2/fileName.json"
and they are both loaded with "$ref":"./fileName.json"  by  "/path1/refferer1.json" , "/path2/refferer2.json"
Currently  the code will try to store them both in jsoneditor.ref  with key  "./fileName.json"
Second one will not be stored at all because such key already exists.
So I think it is better if  store the full url in   the "refs" and refs_with_info" properties

**Comment by pmk65** [comment]
@yaronyam I looked at @desertkun PR, and the changes in` src/editors/select.js` and `src/editors/selectize.js`, enables $ref support in `enumSource` which is a good idea, and should be easy to implement into the current version.
The changes in `src/core.js` is the "fix"

The question is if it's best to make the changes in the` _getExternalRef` function as you have done, or in the `expandRefs` function as @desertkun did?
I haven't worked much with $ref, so I'll let you decide. But personally I think it makes most sense to put it in `expandRefs`, as then you get a schema back with valid references.

**Comment by yaronyam** [comment]
@pmk65 , I'm scratching my head here... it's beed 4 years since I've contributed to this project. Don't recall dealing with external refs functionality. Please double check, perhaps this is relevant for a different contributor

**Comment by pmk65** [comment]
@yaronyam Sorry.I selected the wrong person when I used the @ functionality. The message was meant for @vmanasiev  😁

**Comment by pmk65** [comment]
@vmanasiev To resolve relative urls you can do this (and let the browser do the work 😁):
````javascript
new URL('http://www.example.com/test/', window.location.href).href   
new URL('../mypath', window.location.href).href   
````
or
````javascript
var absPath = function(ref) {
    var link = document.createElement("a");
    link.href = ref;
    return link.href;
}
````

**Comment by Briffou** [comment]
Any news on this ? @vmanasiev Did you submit your code for review ?

**Comment by ericch1** [comment]
I am highly interested by the fix when it will be available!

**Comment by robocoder** [comment]
I think I fixed this in #928
